scvelo/tools/dynamics/_dynamics_base.py

```diff
@@ -0,0 +1,823 @@
+        fit_scaling: Union[bool, int, float] = True,
+        fit_steady_states: bool = True,
+        fit_connected_states: bool = True,
+        fit_basal_transcription=None,
+        high_pars_resolution: bool = False,
+        steady_state_prior: Optional[ndarray] = None,
+        optimization_method: str = "Nelder-Mead",
+    ):
+        # TODO: Check how to remove this
+        self.assignment_mode = None
+
+        self.dynamics = dynamics
+        self.model_parameters = model_parameters
+        self.connectivities = connectivities
+
+        self.fit_basal_transcription = fit_basal_transcription
+        self.fit_connected_states = fit_connected_states
+        self.fit_time = fit_time
+        self.fit_scaling = fit_scaling
+        self.fit_steady_states = fit_steady_states
+        self.high_pars_resolution = high_pars_resolution
+        self.max_iter = max_iter
+        self.percentile = percentile
+        self.steady_state_prior = steady_state_prior
+
+        self.optimization_kwargs = {
+            "method": optimization_method,
+            "options": {"maxiter": int(self.max_iter / 5)},
+        }
+
+    def initialize(
+        self,
+        counts: ndarray,
+        weighted: bool = True,
+        initial_parameter_fit: Optional[Union[Dict, List[Dict]]] = None,
+    ):
+        if initial_parameter_fit is None:
+            initial_parameter_fit = []
+        elif isinstance(initial_parameter_fit, dict):
+            initial_parameter_fit = [initial_parameter_fit]
+
+        counts = counts.copy().astype("float")
+
+        # TODO: Remove attribute counts
+        self.raw_counts = counts.copy()
+
+        weights = (counts > 0).all(axis=1)
+        self.recoverable = weights.sum() > 2
+
+        if not self.recoverable:
+            return
+
+        if weighted:
+            ub = np.percentile(counts[weights, :], self.percentile, axis=0)
+            weights &= (counts[:, ub > 0] <= ub[ub > 0]).all(axis=1)
+
+        self.weights_ = weights.copy()
+        self.std_ = counts[weights, :].std(axis=0)
+        if any(self.std_ == 0):
+            self.std_ = np.ones(len(self.std_))
+
+        self.weights_upper_ = weights
+        if weights.any():
+            self.weights_upper_ &= (counts > (counts[weights, :].max(axis=0) / 3)).all(
+                axis=1
+            )
+
+        weighted_counts = counts[self.weights_, :]
+
+        if isinstance(self.fit_scaling, bool) and self.fit_scaling:
+            self.scaling = self.std_ / self.std_[-1]
+        elif isinstance(self.fit_scaling, bool) and not self.fit_scaling:
+            self.scaling = np.ones(len(self.std_))
+        elif isinstance(self.fit_scaling, (int, float)):
+            self.scaling = np.array([self.fit_scaling, 1])
+            self.fit_scaling = True
+        else:
+            self.scaling = np.array(self.fit_scaling)
+            self.fit_scaling = True
+
+        counts /= self.scaling
+        weighted_counts /= self.scaling
```

Comment:
Yes, I can rename the variable to `counts_subset`. The variable name was taken from the original implementation.
Can you elaborate on the "duplicated division" part? Sorry, not sure what you mean.

Reply:
You could just set `counts_subset = counts[self.weights_, :]`, instead of defining it few lines earlier, and doing the division two times. Just minor..

Reply:
Yes, let's do `counts_subset`.

Reply:
Ah yes, gotcha. Shouldn't we then also call it `self.var_subset` instead of `self.weights`?

Reply:
Yeah, that'd fit better indeed. Wasn't clear at the beginning, whether it's going to be an actual subset or soft weighting - but we're only subsetting.

Reply:
but it's a subset along `obs`.

Reply:
Ah yes, sorry. So `self.obs_subset`.